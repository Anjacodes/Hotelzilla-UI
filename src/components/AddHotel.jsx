import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { postHotelAsync } from '../redux/hotel/hotel-helper';
import { getAllCities } from '../redux/city/city';

const AddHotel = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getAllCities());
  }, []);

  const [newHotel, setNewHotel] = useState({
    name: '',
    description: '',
    image: null,
    rating: 1,
    city: 0,
  });

  const onChangeHandler = (event) => {
    if (event.target.name === 'image') {
      setNewHotel({ ...newHotel, [event.target.name]: event.target.files[0] });
    } else {
      setNewHotel({ ...newHotel, [event.target.name]: event.target.value });
    }
  };

  const navigate = useNavigate();
  const token = useSelector((state) => state.login.token);
  const cities = useSelector((state) => state.city.all);

  const submitHandler = async (event) => {
    event.preventDefault();
    if (newHotel.city !== 0) {
      document.querySelector('#city-error').classList.add('hidden');
      const formData = new FormData();
      formData.append('name', newHotel.name);
      formData.append('description', newHotel.description);
      formData.append('rating', newHotel.rating);
      formData.append('image', newHotel.image);
      formData.append('city_id', newHotel.city);

      await postHotelAsync(token, formData);

      navigate('/', { replace: true });
    } else {
      document.querySelector('#city-error').classList.remove('hidden');
    }
  };

  return (
    <div className="flex w-screen h-screen bg-gray-200 justify-center items-center">
      <form
        onSubmit={submitHandler}
        className="flex flex-col gap-3 items-end p-9 w-sm-3/4"
      >
        <img
          className="self-center w-4/12 w-sm-6/12"
          src="Hotelzilla-logo.png"
          alt=""
        />
        <h2 className="w-full text-center text-2xl">Add a New Hotel</h2>
        <div className="w-full">
          <label htmlFor="name">Name</label>
          <input
            type="text"
            name="name"
            placeholder="Enter hotel name"
            value={newHotel.name}
            onChange={onChangeHandler}
            autoComplete="off"
            className="appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
            required
          />
        </div>
        <div className="w-full">
          <label htmlFor="description">Description</label>
          <textarea
            type="text"
            name="description"
            placeholder="Enter hotel description"
            value={newHotel.description}
            onChange={onChangeHandler}
            autoComplete="off"
            className="appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
            required
          ></textarea>
        </div>
        <div className="w-full">
          <label htmlFor="image">Image</label>
          <input
            type="file"
            name="image"
            accept="image/*"
            multiple={false}
            onChange={onChangeHandler}
            className="appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
            required
          />
        </div>
        <div className="w-full">
          <label htmlFor="rating">Rating</label>
          <input
            type="number"
            name="rating"
            min="1"
            max="5"
            step="0.5"
            value={newHotel.rating}
            onChange={onChangeHandler}
            autoComplete="off"
            className="appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
          />
        </div>
        <div className="w-full">
          <label htmlFor="city">City</label>
          <select
            name="city"
            onChange={onChangeHandler}
            autoComplete="off"
            className="appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
            defaultValue={'defaultSelect'}
            required
          >
            <option value="defaultSelect" disabled>
              Select a city:
            </option>
            {cities.map((city) => {
              return (
                <option key={city.id} value={city.id}>
                  {city.name}
                </option>
              );
            })}
          </select>
          <p id="city-error" className="w-full text-red-600 hidden">
            Please enter a valid city name
          </p>
        </div>
        <button
          type="submit"
          className="bg-lime-400 text-white py-2 appearance-none border rounded px-3 leading-tight focus:outline-none focus:shadow-outline"
        >
          Add
        </button>
      </form>
    </div>
  );
};

export default AddHotel;
