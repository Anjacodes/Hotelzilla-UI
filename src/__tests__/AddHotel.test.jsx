/**
 * @vitest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { it, describe, expect } from 'vitest';
import store from '../redux/configStore';
import AddHotel from '../components/AddHotel';

describe('Add Hotel', () => {
  it('renders correctly', () => {
    const tree = render(
      <Provider store={store}>
        <BrowserRouter>
          <AddHotel />
        </BrowserRouter>
      </Provider>
    );
    expect(tree).toMatchSnapshot();
  });
});
