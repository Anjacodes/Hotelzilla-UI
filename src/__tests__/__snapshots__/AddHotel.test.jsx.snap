// Vitest Snapshot v1

exports[`Add Hotel > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex h-screen items-center justify-center pt-[10vh]"
      >
        <form
          class="w-sm-3/4 flex flex-col items-end gap-3 p-9"
        >
          <img
            alt=""
            class="w-sm-6/12 w-4/12 self-center md:hidden"
            src="Hotelzilla-logo.png"
          />
          <h2
            class="w-full text-center text-3xl text-gray-800"
          >
            Add a New Hotel
          </h2>
          <div
            class="w-full"
          >
            <label
              for="name"
            >
              Name
            </label>
            <input
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="name"
              placeholder="Enter hotel name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="w-full"
          >
            <label
              for="rating"
            >
              Rating
            </label>
            <input
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              max="5"
              min="1"
              name="rating"
              step="0.5"
              type="number"
              value="1"
            />
          </div>
          <div
            class="w-full"
          >
            <label
              for="city"
            >
              Country
            </label>
            <select
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="city"
              required=""
            >
              <option
                disabled=""
                selected=""
                value="defaultSelect"
              >
                Select a country:
              </option>
            </select>
            <p
              class="hidden w-full text-red-600"
              id="city-error"
            >
              Please enter a valid city name
            </p>
          </div>
          <div
            class="w-full"
          >
            <label
              for="description"
            >
              Description
            </label>
            <textarea
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="description"
              placeholder="Enter hotel description"
              required=""
              type="text"
            />
          </div>
          <div
            class="w-full"
          >
            <label
              for="image"
            >
              Image
            </label>
            <input
              accept="image/*"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="image"
              required=""
              type="file"
            />
          </div>
          <button
            class="focus:shadow-outline appearance-none rounded border bg-lime-400 py-2 px-3 leading-tight text-white hover:bg-gray-200 focus:outline-none"
            type="submit"
          >
            Add
          </button>
        </form>
        <i
          class="fa-solid fa-caret-left absolute left-0 bottom-4 rounded-r-full bg-lime-400 py-4 px-6 text-slate-50 hover:bg-lime-500 md:hidden"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex h-screen items-center justify-center pt-[10vh]"
    >
      <form
        class="w-sm-3/4 flex flex-col items-end gap-3 p-9"
      >
        <img
          alt=""
          class="w-sm-6/12 w-4/12 self-center md:hidden"
          src="Hotelzilla-logo.png"
        />
        <h2
          class="w-full text-center text-3xl text-gray-800"
        >
          Add a New Hotel
        </h2>
        <div
          class="w-full"
        >
          <label
            for="name"
          >
            Name
          </label>
          <input
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="name"
            placeholder="Enter hotel name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="w-full"
        >
          <label
            for="rating"
          >
            Rating
          </label>
          <input
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            max="5"
            min="1"
            name="rating"
            step="0.5"
            type="number"
            value="1"
          />
        </div>
        <div
          class="w-full"
        >
          <label
            for="city"
          >
            Country
          </label>
          <select
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="city"
            required=""
          >
            <option
              disabled=""
              selected=""
              value="defaultSelect"
            >
              Select a country:
            </option>
          </select>
          <p
            class="hidden w-full text-red-600"
            id="city-error"
          >
            Please enter a valid city name
          </p>
        </div>
        <div
          class="w-full"
        >
          <label
            for="description"
          >
            Description
          </label>
          <textarea
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="description"
            placeholder="Enter hotel description"
            required=""
            type="text"
          />
        </div>
        <div
          class="w-full"
        >
          <label
            for="image"
          >
            Image
          </label>
          <input
            accept="image/*"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="image"
            required=""
            type="file"
          />
        </div>
        <button
          class="focus:shadow-outline appearance-none rounded border bg-lime-400 py-2 px-3 leading-tight text-white hover:bg-gray-200 focus:outline-none"
          type="submit"
        >
          Add
        </button>
      </form>
      <i
        class="fa-solid fa-caret-left absolute left-0 bottom-4 rounded-r-full bg-lime-400 py-4 px-6 text-slate-50 hover:bg-lime-500 md:hidden"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex h-screen items-center justify-center pt-[10vh]"
      >
        <form
          class="w-sm-3/4 flex flex-col items-end gap-3 p-9"
        >
          <img
            alt=""
            class="w-sm-6/12 w-4/12 self-center md:hidden"
            src="Hotelzilla-logo.png"
          />
          <h2
            class="w-full text-center text-3xl text-gray-800"
          >
            Add a New Hotel
          </h2>
          <div
            class="w-full"
          >
            <label
              for="name"
            >
              Name
            </label>
            <input
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="name"
              placeholder="Enter hotel name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="w-full"
          >
            <label
              for="rating"
            >
              Rating
            </label>
            <input
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              max="5"
              min="1"
              name="rating"
              step="0.5"
              type="number"
              value="1"
            />
          </div>
          <div
            class="w-full"
          >
            <label
              for="city"
            >
              Country
            </label>
            <select
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="city"
              required=""
            >
              <option
                disabled=""
                selected=""
                value="defaultSelect"
              >
                Select a country:
              </option>
            </select>
            <p
              class="hidden w-full text-red-600"
              id="city-error"
            >
              Please enter a valid city name
            </p>
          </div>
          <div
            class="w-full"
          >
            <label
              for="description"
            >
              Description
            </label>
            <textarea
              autocomplete="off"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="description"
              placeholder="Enter hotel description"
              required=""
              type="text"
            />
          </div>
          <div
            class="w-full"
          >
            <label
              for="image"
            >
              Image
            </label>
            <input
              accept="image/*"
              class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
              name="image"
              required=""
              type="file"
            />
          </div>
          <button
            class="focus:shadow-outline appearance-none rounded border bg-lime-400 py-2 px-3 leading-tight text-white hover:bg-gray-200 focus:outline-none"
            type="submit"
          >
            Add
          </button>
        </form>
        <i
          class="fa-solid fa-caret-left absolute left-0 bottom-4 rounded-r-full bg-lime-400 py-4 px-6 text-slate-50 hover:bg-lime-500 md:hidden"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex h-screen items-center justify-center pt-[10vh]"
    >
      <form
        class="w-sm-3/4 flex flex-col items-end gap-3 p-9"
      >
        <img
          alt=""
          class="w-sm-6/12 w-4/12 self-center md:hidden"
          src="Hotelzilla-logo.png"
        />
        <h2
          class="w-full text-center text-3xl text-gray-800"
        >
          Add a New Hotel
        </h2>
        <div
          class="w-full"
        >
          <label
            for="name"
          >
            Name
          </label>
          <input
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="name"
            placeholder="Enter hotel name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="w-full"
        >
          <label
            for="rating"
          >
            Rating
          </label>
          <input
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            max="5"
            min="1"
            name="rating"
            step="0.5"
            type="number"
            value="1"
          />
        </div>
        <div
          class="w-full"
        >
          <label
            for="city"
          >
            Country
          </label>
          <select
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="city"
            required=""
          >
            <option
              disabled=""
              selected=""
              value="defaultSelect"
            >
              Select a country:
            </option>
          </select>
          <p
            class="hidden w-full text-red-600"
            id="city-error"
          >
            Please enter a valid city name
          </p>
        </div>
        <div
          class="w-full"
        >
          <label
            for="description"
          >
            Description
          </label>
          <textarea
            autocomplete="off"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="description"
            placeholder="Enter hotel description"
            required=""
            type="text"
          />
        </div>
        <div
          class="w-full"
        >
          <label
            for="image"
          >
            Image
          </label>
          <input
            accept="image/*"
            class="focus:shadow-outline w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-700 focus:outline-none"
            name="image"
            required=""
            type="file"
          />
        </div>
        <button
          class="focus:shadow-outline appearance-none rounded border bg-lime-400 py-2 px-3 leading-tight text-white hover:bg-gray-200 focus:outline-none"
          type="submit"
        >
          Add
        </button>
      </form>
      <i
        class="fa-solid fa-caret-left absolute left-0 bottom-4 rounded-r-full bg-lime-400 py-4 px-6 text-slate-50 hover:bg-lime-500 md:hidden"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
